modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java

```diff
@@ -0,0 +1,449 @@
+                // Unbounded search, include all.
+                if (boundary == null)
+                    return false;
+
+                int criteriaKeysCnt = treeCriteria.size();
+
+                for (int i = 0; i < criteriaKeysCnt; i++) {
+                    RangeIndexQueryCriterion c = treeCriteria.get(i);
+
+                    // Include all values on this field.
+                    if (boundary.key(i) == null)
+                        continue;
+
+                    int cmp = rowCmp.compareKey(row, boundary, i);
+
+                    // Swap direction.
+                    if (hnd.indexKeyDefinitions().get(i).order().sortOrder() == DESC)
+                        cmp = -1 * cmp;
+
+                    // Exclude if field equals boundary field and criteria is excluding.
+                    if (cmp == 0) {
+                        if (boundarySign > 0 && !c.lowerIncl())
+                            return true;
+
+                        if (boundarySign < 0 && !c.upperIncl())
+                            return true;
+                    }
+
+                    // Check sign. Exclude if field is out of boundaries.
+                    if (cmp * boundarySign < 0)
+                        return true;
+                }
+
+                return false;
+            }
+        };
+    }
+
+    /** */
+    private IndexKey key(Object val, IndexKeyDefinition def, IndexKeyTypeSettings settings, CacheObjectContext coctx) {
+        IndexKey key = null;
+
+        if (val != null) {
+            if (val instanceof IndexQueryCriteriaBuilder.Null)
+                val = null;
+
+            key = IndexKeyFactory.wrap(
+                val, def.idxType(), coctx, settings);
+        }
+
+        return key;
+    }
+
+    /** Single cursor over multiple segments. Next value is choose with the index row comparator. */
+    private class SegmentedIndexCursor implements GridCursor<IndexRow> {
+        /** Cursors over segments. */
+        private final GridCursor<IndexRow>[] cursors;
+
+        /** Whether returns first value for user. */
+        private boolean returnFirst;
+
+        /** Offset of current segmented cursor to return value. */
+        private int cursorOff;
+
+        /** Comparator to compare index rows. */
+        private final Comparator<GridCursor<IndexRow>> cursorComp;
+
+        /** */
+        SegmentedIndexCursor(GridCursor<IndexRow>[] cursors, IndexRowComparator rowCmp) {
+            this.cursors = cursors;
+
+            cursorComp = new Comparator<GridCursor<IndexRow>>() {
+                @Override public int compare(GridCursor<IndexRow> o1, GridCursor<IndexRow> o2) {
+                    try {
+                        if (o1 == o2)
+                            return 0;
+
+                        if (o1 == null)
+                            return -1;
+
+                        if (o2 == null)
+                            return 1;
```

Comment:
Does 'nulls' order independent from index sort order?

Reply:
Those nulls is not about rows but about cursors (if some cursor is NULL then we compare it this way). But actually this check is useless as there cant be nulls at all. Will remove this.